### ISSUE TITLE
Backoffs should have jitter

### ISSUE THREAD

**Issue by vllry** [kubernetes/kubernetes#87915]
client-go implements exponential backoff. This backoff is used in the control plane, kubelets, and by many 3rd parties. There are many cases where backoff would be improved by jitter.

For example, suppose many pods enter a CrashLoopBackOff at once, due to a shared failure.
With no jitter, this (potentially large) group of pods will all retry and start at the same time, hammering the Kubernetes API and other systems. In addition, recovery would be delayed up to 5 minutes (the max backoff in that context) until all pods start at the same time.

**Comment by lavalamp** [comment]
https://github.com/kubernetes/kubernetes/pull/87829 just went in-- related.

**Comment by vllry** [comment]
> #87829 just went in-- related.

Naively, why does the api machinery have a separate implementation from client-go backoff?

**Comment by lavalamp** [comment]
I support client-go tweaks for this purpose, but as time goes on we will accumulate more and more non-client-go clients. So the solution here is really to make the control plane more robust.

This has been in flight for a while: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md

For the specific case you mention, if the system is configured right, the clients will end up grouped in a few queues, interfering with each other but not with other api callers. Waiting in a queue rather than failing immediately (current max-in-flight limit behavior) also solves the problem of clients going into a hot loop and retrying very rapidly.

We currently expect a usable alpha in 1.18.

**Comment by lavalamp** [comment]
https://github.com/kubernetes/kubernetes/issues/87794 is the thing I actually should have linked as relevant.

As for why it's divergent, I think the client go backoff just isn't in the right shape for our collection of WaitXXX methods; also apimachinery can't import the client.

**Comment by vllry** [comment]
> I support client-go tweaks for this purpose, but as time goes on we will accumulate more and more non-client-go clients. So the solution here is really to make the control plane more robust.
> 
> This has been in flight for a while: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
> 
> For the specific case you mention, if the system is configured right, the clients will end up grouped in a few queues, interfering with each other but not with other api callers. Waiting in a queue rather than failing immediately (current max-in-flight limit behavior) also solves the problem of clients going into a hot loop and retrying very rapidly.
> 
> We currently expect a usable alpha in 1.18.

This seems to address a subset of concerns. Mass pod startup can cause strain outside of the apiserver (EG IO-heavy bootstrapping for the app). Additionally, it doesn't look like this addresses the lock-step nature of unjittered backoffs (which still cause an artificial recovery delay, aside from the load concerns in a thundering herd). Not sure if I'm missing something here - these look like complimentary solutions to me.

**Comment by lavalamp** [comment]
> This seems to address a subset of concerns. Mass pod startup can cause strain outside of the apiserver (EG IO-heavy bootstrapping for the app). 

What is the scenario in which a client go jitter changes mass pod *start* behavior? I agree that is a potential problem (assuming I understand you) but I don't see the connection, I'd expect e.g. jittering watch deliveries as more targeted at that.

> Additionally, it doesn't look like this addresses the lock-step nature of unjittered backoffs (which still cause an artificial recovery delay, 

Maybe-- everyone will be sitting in a queue for different lengths of time, so it should help somewhat. (we could go even further and add on the time spent waiting for the network to the delay for the next round, but a jitter is probably easier.)

That's assuming that apiserver is healthy enough to put that many connections on hold; we know apiserver can do it in steady state because the vast majority of these clients are watching stuff, the question is mostly about the seconds after it (tries to) start.

> aside from the load concerns in a thundering herd). Not sure if I'm missing something here - these look like complimentary solutions to me.

Yes, they are complementary. My point is more that we, the control plane, need to get our house in order before we start scolding clients too much about their behavior :) I really think the control plane can and should improve 10x.

**Comment by lavalamp** [comment]
https://github.com/kubernetes/kubernetes/pull/87795 addresses list/watch calls, and there's still plenty of room to address all calls with your change (which I left comments on).

**Comment by fedebongio** [comment]
Maybe an item to discuss live in SIG meeting?
